demonstrations/function_fitting_qsp.py

```diff
@@ -0,0 +1,922 @@
+r"""
+
+.. function_fitting_qsp:
+
+Function Fitting using Quantum Signal Processing (QSP)
+======================================================
+
+.. meta::
+    :property="og:description": Learn how to create polynomial approximations to functions
+        using Quantum Signal Processing (QSP).
+    :property="og:image": https://pennylane.ai/qml/demonstrations/function_fitting_qsp/cover.png
+
+.. figure:: ../demonstrations/function_fitting_qsp/trained_poly.gif
+    :align: center
+    :width: 50%
+
+1. Introduction:
+~~~~~~~~~~~~~~~~
+
+This demo was inspired by the paper `‘A Grand Unification of Quantum
+Algorithms’ <https://arxiv.org/abs/2105.02859>`__; this may sound like a
+very ambitious title, but in my opinion the authors fully deliver. This
+paper is centered around the ‘Quantum Singular Value Transform’ (QSVT)
+protocol and how it provides a single framework to generalize some of
+the most famous quantum algorithms like Shor’s factoring, Grover search,
+and more.
+
+The QSVT is a method to apply polynomial transformations to the singular
+values of *any matrix* (does not have to be square). This is powerful
+because from polynomial transformations we can generate arbitrary function
+transformations (using taylor approximations). The QSVT protocol is an
+extension of the more constrained ‘Quantum Signal Processing’ (QSP)
+protocol which presents a method for polynomial transformation of matrix
+entries in a single-qubit unitary operator. The QSVT protocol is quite
+sophisticated and involved in its derivation, but the idea at its core
+is quite simple. By studying QSP, we get a relatively simpler path to explore
+this idea at the foundation of QSVT.
+
+In this demo, we will explore the QSP protocol and how it can be used
+for curve fitting. This is a powerful tool that will ultimately allow us
+to approximate any function on the interval :math:`[-1, 1]` that
+satisfies certain constraints. Before we can dive into function fitting,
+let’s develop some intuition. Consider the following operator
+parameterized by :math:`a \in [-1, 1]`:
+
+.. math:: \hat{W}(a) = \begin{bmatrix} a & i\sqrt{1 - a^{2}} \\ i\sqrt{1 - a^{2}} & a \end{bmatrix}
+
+:math:`\hat{W}(a)` is called the *Signal Rotation Operator* (SRO). In
+this particular case we are rotating around the x-axis but it can, in
+general, take other forms. Using this operator, we can construct another
+operator called the *Signal Processing Operator* (SPO),
+
+.. math::  \hat{U}_{sp} = \hat{R}_{z}(\phi_{0}) \prod_{k=1}^{d} \hat{W}(a) \hat{R}_{z}(\phi_{k})
```

Comment:
```suggestion
.. math::  \hat{U}_{sp} = \hat{R}_{z}(\phi_{0}) \prod_{k=1}^{d} \hat{W}(a) \hat{R}_{z}(\phi_{k}).
```
Equations are part of the text too and so they require punctuation